crates/sui-types/src/programmable_transaction_builder.rs

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Utility for generating programmable transactions, either by specifying a command or for
+//! migrating legacy transactions
+
+use indexmap::IndexSet;
+use move_core_types::{identifier::Identifier, language_storage::TypeTag};
+use serde::Serialize;
+
+use crate::{
+    base_types::{ObjectID, ObjectRef, SuiAddress},
+    messages::{
+        Argument, CallArg, Command, MoveCall, MoveModulePublish, ObjectArg, Pay, PayAllSui, PaySui,
+        ProgrammableMoveCall, ProgrammableTransaction, SingleTransactionKind, TransactionData,
+        TransactionKind, TransferObject, TransferSui,
+    },
+};
+
+pub fn migrate_transaction_data(mut m: TransactionData) -> anyhow::Result<TransactionData> {
+    let mut builder = ProgrammableTransactionBuilder::new();
+    match m.kind {
+        TransactionKind::Single(SingleTransactionKind::PaySui(PaySui {
+            coins: _coins,
+            recipients,
+            amounts,
+        })) => {
+            builder.pay_sui(recipients, amounts)?;
+            anyhow::bail!("blocked by gas smashing")
+        }
+        TransactionKind::Single(SingleTransactionKind::PayAllSui(PayAllSui {
+            coins: _coins,
+            recipient,
+        })) => {
+            builder.pay_all_sui(recipient);
+            anyhow::bail!("blocked by gas smashing")
+        }
+        TransactionKind::Single(t) => builder.single_transaction(t)?,
+        TransactionKind::Batch(ts) => {
+            for t in ts {
+                builder.single_transaction(t)?
+            }
+        }
+    };
+    let pt = builder.finish();
+    m.kind = TransactionKind::Single(SingleTransactionKind::ProgrammableTransaction(pt));
+    Ok(m)
+}
+
+#[derive(Default)]
+pub struct ProgrammableTransactionBuilder {
+    inputs: IndexSet<CallArg>,
+    commands: Vec<Command>,
+}
+
+impl ProgrammableTransactionBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn finish(self) -> ProgrammableTransaction {
+        let Self { inputs, commands } = self;
+        let inputs = inputs.into_iter().collect();
+        ProgrammableTransaction { inputs, commands }
+    }
+
+    pub fn pure<T: Serialize>(&mut self, value: T) -> anyhow::Result<Argument> {
+        Ok(self.input(CallArg::Pure(bcs::to_bytes(&value).map_err(|e| {
+            anyhow::anyhow!("Unable to serialize pure argument. Got error: {e}")
+        })?)))
```

Comment:
`anyhow` has a `context` feature that allows you to do this slightly more cleanly:

```suggestion
        Ok(self.input(CallArg::Pure(bcs::to_bytes(&value).context("Serializing pure argument")?)))
```